python/glow/gwas/linear_regression.py

```diff
@@ -72,31 +80,79 @@ def linear_regression(genotype_df: DataFrame,
         C = _add_intercept(C, phenotype_df.shape[0])
 
     # Prepare covariate basis and phenotype residuals
+    Q = np.linalg.qr(C)[0]
     Y = phenotype_df.to_numpy(np.float64, copy=True)
     Y_mask = (~np.isnan(Y)).astype(np.float64)
     np.nan_to_num(Y, copy=False)
-    Q = np.linalg.qr(C)[0]
-    Y = _residualize_in_place(Y, Q) * Y_mask
-    YdotY = np.sum(Y * Y, axis=0)
+    _residualize_in_place(Y, Q)
+
+    if not offset_df.empty:
+        if not _indexes_have_same_elements(phenotype_df.columns, offset_df.columns):
+            raise ValueError(f'phenotype_df and offset_df should have the same column names.')
+        if offset_df.index.nlevels == 1:  # Indexed by sample id
+            if not _indexes_have_same_elements(phenotype_df.index, offset_df.index):
+                raise ValueError(f'phenotype_df and offset_df should have the same index.')
+            Y_state = _create_YState_from_offset(Y, Y_mask, phenotype_df, offset_df)
+        elif offset_df.index.nlevels == 2:  # Indexed by sample id and contig
+            all_contigs = offset_df.index.get_level_values(1).unique()
+            Y_state = {}
+            for contig in all_contigs:
+                offset_for_contig = offset_df.xs(contig, level=1)
+                if not _indexes_have_same_elements(phenotype_df.index, offset_for_contig.index):
+                    raise ValueError(
+                        'When using a multi-indexed offset_df, the offsets for each contig '
+                        'should have the same index as phenotype_df')
+                Y_state[contig] = _create_YState_from_offset(Y, Y_mask, phenotype_df,
+                                                             offset_for_contig)
+    else:
+        Y_state = _create_YState(Y, Y_mask)
+
     dof = C.shape[0] - C.shape[1] - 1
 
     def map_func(pdf_iterator):
         for pdf in pdf_iterator:
-            yield _linear_regression_inner(pdf, Y, YdotY, Y_mask, Q, dof,
-                                           phenotype_df.columns.to_series().astype('str'),
-                                           values_column)
+            yield _linear_regression_dispatch(pdf, Y_state, Y_mask, Q, dof,
+                                              phenotype_df.columns.to_series().astype('str'),
+                                              values_column)
 
     return genotype_df.mapInPandas(map_func, result_struct)
 
 
+def _indexes_have_same_elements(idx1: pd.Index, idx2: pd.Index) -> bool:
```

Comment:
perhaps not call this function `_indexes...` as you use it for non-index comparisons.